### PR TITLE
Throw IllegalArgumentException when jib.to.tags is set to a collection containing a null value. 

### LINKED ISSUE #2760: NullPointerException when `jib.to.tags` is set to Set<String> containing `null`
**Description of the issue**:
NullPointerException thrown when `jib.to.tags`is set to a `Set<String>` containing `null`.  For example, `jib.to.tags= ["tag1", null] as Set` in `build.gradle`. 

**Expected behavior**:
Don't throw NullPointerException. 

**Log output**: 
```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':jib'.
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:38)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:49)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:315)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:305)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:101)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:49)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.NullPointerException
        at org.gradle.api.internal.provider.ValidatingValueCollector.add(ValidatingValueCollector.java:37)
        at org.gradle.api.internal.provider.ValidatingValueCollector.addAll(ValidatingValueCollector.java:45)
        at org.gradle.api.internal.provider.Collectors$ElementsFromCollection.maybeCollectInto(Collectors.java:189)
        at org.gradle.api.internal.provider.AbstractCollectionProperty.doGetOrNull(AbstractCollectionProperty.java:168)
        at org.gradle.api.internal.provider.AbstractCollectionProperty.getOrNull(AbstractCollectionProperty.java:162)
        at org.gradle.api.internal.provider.AbstractCollectionProperty.getOrNull(AbstractCollectionProperty.java:37)
        at com.google.cloud.tools.jib.gradle.TargetImageParameters.getTags(TargetImageParameters.java:78)
        at com.google.cloud.tools.jib.gradle.TargetImageParameters_Decorated.getTags(Unknown Source)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue$1$1.create(AbstractNestedRuntimeBeanNode.java:82)
        at org.gradle.util.SingleMessageLogger.whileDisabled(SingleMessageLogger.java:341)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue$1.get(AbstractNestedRuntimeBeanNode.java:79)
        at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:167)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue.getValue(AbstractNestedRuntimeBeanNode.java:145)
        at org.gradle.api.internal.tasks.properties.bean.AbstractNestedRuntimeBeanNode$DefaultPropertyValue.validate(AbstractNestedRuntimeBeanNode.java:163)
        at org.gradle.api.internal.tasks.DefaultTaskInputPropertySpec.validate(DefaultTaskInputPropertySpec.java:68)
        at org.gradle.api.internal.tasks.execution.DefaultTaskProperties.validate(DefaultTaskProperties.java:197)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:49)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:119)
        at org.gradle.api.internal.tasks.execution.ResolvePreviousStateExecuter.execute(ResolvePreviousStateExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:93)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:45)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:94)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        ... 22 more
```

### PR DESCRIPTION
**Pull request by mpeddada1**
Fixes #2760

### CODE REVIEW

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java [hunk 1]
```diff
@@ -60,10 +61,16 @@ public void setImage(String image) {
   @Input
   @Optional
   public Set<String> getTags() {
-    if (System.getProperty(PropertyNames.TO_TAGS) != null) {
-      return ImmutableSet.copyOf(
-          ConfigurationPropertyValidator.parseListProperty(
-              System.getProperty(PropertyNames.TO_TAGS)));
+    String property = System.getProperty(PropertyNames.TO_TAGS);
+    Set<String> tags =
```
Comment (loosebazooka) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:65]: For readability: I think maybe `tags` is confusing with the class member `tags` (`this.tags`). Perhaps use something that differentiates it? Personally I don't think `this.tags` should be used very often outside constructors.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java [hunk 2]
```diff
@@ -60,10 +61,16 @@ public void setImage(String image) {
...
+        property != null
+            ? ImmutableSet.copyOf(
+                ConfigurationPropertyValidator.parseListProperty(
+                    System.getProperty(PropertyNames.TO_TAGS)))
+            : this.tags;
+    if (tags.stream().anyMatch(Strings::isNullOrEmpty)) {
```
Comment (loosebazooka) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:71]: should we handle whitespace?

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java [hunk 3]
```diff
@@ -85,6 +86,21 @@ public void testTo() {
     Assert.assertEquals("some cred helper", testJibExtension.getTo().getCredHelper());
     Assert.assertEquals("some username", testJibExtension.getTo().getAuth().getUsername());
     Assert.assertEquals("some password", testJibExtension.getTo().getAuth().getPassword());
+
+    // Testing scenario where jib.to.tags is set to a collection containing a null value.
+    HashSet<String> tags = new HashSet<String>();
+    tags.add(null);
```
Comment (loosebazooka) [inline jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java:92]: if we fail on empty, we should probably add a test for empty as well

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java [hunk 4]
```diff
@@ -85,6 +86,21 @@ public void testTo() {
     Assert.assertEquals("some cred helper", testJibExtension.getTo().getCredHelper());
     Assert.assertEquals("some username", testJibExtension.getTo().getAuth().getUsername());
     Assert.assertEquals("some password", testJibExtension.getTo().getAuth().getPassword());
+
+    // Testing scenario where jib.to.tags is set to a collection containing a null value.
```
Comment (loosebazooka) [inline jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java:90]: Maybe just put this in a new test?

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java (hunk 3 above)
Comment (chanseokoh) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java:92]: Also the case of getting tags from the system property.

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java [hunk 5]
```diff
@@ -85,6 +86,21 @@ public void testTo() {
...
+    HashSet<String> tags = new HashSet<String>();
+    tags.add(null);
+    tags.add("tags1");
+    try {
+      testJibExtension.to(
+          to -> {
+            to.setTags(tags);
+          });
+      testJibExtension.getTo().getTags();
+      Assert.fail("Expect this to fail");
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java:100]: Nit: if the message doesn't provide additional helpful information or details, just use `Assert.fail()`.

And I think it's meaningful to move `testJibExtension.to()` out of the try-catch block.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java [hunk 6]
```diff
@@ -60,10 +61,16 @@ public void setImage(String image) {
...
-    if (System.getProperty(PropertyNames.TO_TAGS) != null) {
-      return ImmutableSet.copyOf(
-          ConfigurationPropertyValidator.parseListProperty(
-              System.getProperty(PropertyNames.TO_TAGS)));
+    String property = System.getProperty(PropertyNames.TO_TAGS);
+    Set<String> tags =
+        property != null
+            ? ImmutableSet.copyOf(
+                ConfigurationPropertyValidator.parseListProperty(
+                    System.getProperty(PropertyNames.TO_TAGS)))
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:69]:
```suggestion
                    property))
```

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java (hunk 2 above)
Comment (mpeddada1) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java]: Oh that didn't occur to me. Thanks!

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java (hunk 3 above)
Comment (mpeddada1) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java]: good catch!

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java (hunk 5 above)
Comment (mpeddada1) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibExtensionTest.java]: Ah yes, having all the extra lines outside the try-catch makes sense. Thanks!

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java (hunk 6 above)
Comment (mpeddada1) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java]: Done:)

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java [hunk 7]
```diff
@@ -60,12 +61,19 @@ public void setImage(String image) {
...
+    Set<String> tagsValue =
+        property != null
+            ? ImmutableSet.copyOf(ConfigurationPropertyValidator.parseListProperty(property))
+            : tags;
+    String source = property != null ? PropertyNames.TO_TAGS : "jib.to.tags";
+    if (tagsValue.stream().anyMatch(Strings::isNullOrEmpty)) {
+      throw new IllegalArgumentException(source + " has empty tag");
     }
-    return tags;
+    if (tagsValue.stream().anyMatch((String str) -> str.contains(" "))) {
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:73]:
```suggestion
    if (tagsValue.stream().anyMatch(str -> str.contains(" "))) {
```

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java [hunk 8]
```diff
@@ -60,12 +61,19 @@ public void setImage(String image) {
...
+    Set<String> tagsValue =
+        property != null
+            ? ImmutableSet.copyOf(ConfigurationPropertyValidator.parseListProperty(property))
+            : tags;
+    String source = property != null ? PropertyNames.TO_TAGS : "jib.to.tags";
+    if (tagsValue.stream().anyMatch(Strings::isNullOrEmpty)) {
+      throw new IllegalArgumentException(source + " has empty tag");
     }
-    return tags;
+    if (tagsValue.stream().anyMatch(str -> str.contains(" "))) {
```
Comment (loosebazooka) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:73]: What I meant was checking for something like `"   "` but maybe jib-core already validates these values before going?
Perhaps your original condition was enough, lemme see if jib core validates this.

Comment (loosebazooka) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:73]: Sorry, yeah so looks like we handle this properly already at the processing level: https://github.com/GoogleContainerTools/jib/blob/master/jib-core/src/main/java/com/google/cloud/tools/jib/api/ImageReference.java#L68
so `isNullOrEmpty` should be sufficient?

**Review by loosebazooka — APPROVED** [review]

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java [hunk 9]
```diff
@@ -69,12 +69,21 @@ public void setImage(Provider<String> image) {
...
+    } else {
+      try {
+        tagsValue = tags.get();
+      } catch (NullPointerException ex) {
+        throw new IllegalArgumentException("jib.to.tags contains null tag");
+      }
     }
-    return tags.get();
+    if (tagsValue.stream().anyMatch(str -> str.isEmpty())) {
+      throw new IllegalArgumentException("jib.to.tags contains empty tag");
```
Comment (loosebazooka) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:84]: Do we have to do that logic where we know its from system property or from configuration in teh build file?

Comment (mpeddada1) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/TargetImageParameters.java:84]: In terms of error messaging, I didn't think having that logic added much value since regardless of whether it was from system property or configuration, (in `String source = property != null ? PropertyNames.TO_TAGS : "jib.to.tags";`) the message would still remain the same?

**Review by loosebazooka — APPROVED** [review]

**Review by mpeddada1 — COMMENTED** [review]
Added this issue to the CHANGELOG.
